### ISSUE TITLE
Question on Chapter 10: "ap will never change container types"

### ISSUE THREAD

**Issue by thurt** [MostlyAdequate/mostly-adequate-guide#272]
> you should know that applicatives are "closed under composition", meaning ap will never change container types on us(yet another reason to favor over monads)
```javascript
Container.prototype.ap = function(other_container) {
  return other_container.map(this.__value);
}
```

According to the provided definition of `ap`, it appears that you _could_ change container types since there is no type checking on the two containers. I see the same issue with the `lift2A` function. 

Wouldn't this example change the initial Container type to the Maybe type?:
```javascript
Container.of((x) => console.log(x)).ap(Maybe.of(2))
```

**Comment by thurt** [comment]
Or consider `liftA2`:
```javascript
var liftA2 = function(f, functor1, functor2) {
  return functor1.map(f).ap(functor2)
}

//  two different types -- the .ap in liftA2 changes the type from Container to Maybe
liftA2(add, Container.of(2), Maybe.of(3))
```

**Comment by safareli** [comment]
If you look at definition of ap from [here](https://hackage.haskell.org/package/base-4.8.2.0/docs/Control-Applicative.html#v:-60--42--62-) `(<*>) :: f (a -> b) -> f a -> f b` it shows that all functors should be of same type. 

So yes type will change as we are not type checking it in js, but I think you should not do that.

**Comment by thurt** [comment]
Thanks @safareli -- this is what I suspected. 

I think there should be another sentence in the book on this topic of type checking, like "In JS, you must check that you are using the same functor type on both sides of `ap`. Otherwise, you end up with a version of `ap` that is inconsistent with its definition in other FP languages."

**Comment by safareli** [comment]
That looks good to me

**Comment by KtorZ** [comment]
Agreed that this deserve a paragraph in the book somewhere in the introduction or preface. A lot of functions and data structures presented actually holds under the assumption that they're used in the way they're expected to be used. 

Will do.
